src/handshake/server.rs

```diff
@@ -33,27 +34,90 @@ const SOKETTO_VERSION: &str = env!("CARGO_PKG_VERSION");
 
 /// Websocket handshake client.
 #[derive(Debug)]
-pub struct Server<'a, T> {
-    socket: T,
+pub struct Server<'a, Socket, Hosts = AllowAny, Origins = AllowAny> {
+    socket: Socket,
     /// Protocols the server supports.
     protocols: Vec<&'a str>,
     /// Extensions the server supports.
     extensions: Vec<Box<dyn Extension + Send>>,
     /// Encoding/decoding buffer.
-    buffer: BytesMut
+    buffer: BytesMut,
+    /// Check to be performed on `Host` header during the handshake.
+    hosts: Hosts,
+    /// Check to be performed on `Origin` header during the handshake.
+    origins: Origins,
 }
 
-impl<'a, T: AsyncRead + AsyncWrite + Unpin> Server<'a, T> {
+impl<'a, Socket> Server<'a, Socket>
+where
+    Socket: AsyncRead + AsyncWrite + Unpin,
+{
     /// Create a new server handshake.
-    pub fn new(socket: T) -> Self {
+    pub fn new(socket: Socket) -> Self {
         Server {
             socket,
             protocols: Vec::new(),
             extensions: Vec::new(),
-            buffer: BytesMut::new()
+            buffer: BytesMut::new(),
+            hosts: AllowAny,
+            origins: AllowAny,
         }
     }
+}
+
+impl<'a, Socket, Hosts, Origins> Server<'a, Socket, Hosts, Origins> {
+    /// Set the allowed `Host` headers.
+    pub fn set_hosts<NewHosts>(self, hosts: NewHosts) -> Server<'a, Socket, NewHosts, Origins>
```

Comment:
ok, why `self` instead of `&mut self` and just replace self.hosts?

Reply:
Because they are different types I think, you feed it a `Server<'a, Socket, Hosts, Origins>` and get back a `Server<'a, Socket, DifferentKindOfHosts, Origins>`.

Reply:
Because the type of `self.hosts` is different.